File: 0150-Eval-RPN/0150.py
class Solution:
  def evalRPN(self, tokens: List[str]) -> int:
    op_to_binary_fn = {
      '+': add,
      '-': sub,
      '*': mul,
      '/': lambda x, y: int(x / y),
    }

    stack = list()
    for token in tokens:
      try:
        num = int(token)
      except ValueError:
        num2 = stack.pop()
        num1 = stack.pop()
        num = op_to_binary_fn[token](num1, num2)
      finally:
        stack.append(num)

    return stack[0]

File: 101-200/0150-Eval-RPN/0150.java
class Solution {
  public int evalRPN(String[] tokens) {
    Deque<Integer> stack = new LinkedList<Integer>();
    int n = tokens.length;
    for (int i = 0; i < n; i++) {
      String token = tokens[i];
      if (isNumber(token)) {
        stack.push(Integer.parseInt(token));
      } else {
        int num2 = stack.pop();
        int num1 = stack.pop();
        switch (token) {
          case '+':
            stack.push(num1 + num2);
            break;
          case '-':
            stack.push(num1 - num2);;
            break;
          case '*':
            stack.push(num1 * num2);
            break;
          case '/':
            stack.push(num1 / num2);
            break;
          default:
            break;
        }
      }
    }
    return stack.pop();
  }

  public boolean isNumber(String token) {
    return !('+'.equals(token) || '-'.equals(token) || '*'.equals(token) || '/'.equals(token));
  }
}